enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,652 @@
+and a more sophisticated route health detection. That means that the underlying
+networking solution (OVS in the case of both OVN-Kubernetes and openshift-sdn)
+will manage the route health check. It is worth noting that the "amount of
+re-balancing" done by the network plugin is expected to be reduced, since the
+network fabric will be able to load balance traffic between all egress nodes
+provided and always use a healthy node (thus not requiring the network plugin to
+perform an egress IP re-assignment every time a route goes down).
+
+Given the above, it is however important that the network plugin is conveyed the
+cloud's information concerning the nodes' subnet range. This is especially
+evident on GCP, as without it: the network plugin won't be able to perform any
+assignments at all, given the nodes' default `/32` subnet.
+
+#### Inter-component communication
+
+Multiple pieces of information needs to be communicated between the network
+plugins and this cloud component:
+
+1. The cloud's node subnet and capacity information.
+2. The egress IP assignment to be performed by the cloud component
+3. The status of the egress IP assignment performed by the cloud component
+
+Two models have been decided for this: a CRD based approach and a node
+annotated approach.
+
+##### annotated approach
+
+As to satisfy condition 1. the cloud component will annotate all cluster nodes
+with the nodes' cloud subnet as retrieved from the cloud provider's API when it
+is initialized. This annotation will have the following specification:
+
+`cloud.openshift.io/private-ipconfig: {"ifaddr": {"ipv4": "$IPv4_ADDRESS/$IPv4_SUBNET_MASK", "ipv6": "$IPv6_ADDRESS/$IPv6_SUBNET_MASK"}, "capacity": {"ipv4": "$IPv4_CAPACITY",  "ipv6": "$IPv6_CAPACITY"}}`
+
+This will in turn be read by the network plugin and used during the egress IP
+assignment, superseding any host level subnet information already parsed by the
+network plugin and conveying the capacity to the network plugin's assignment
+algorithm.
+
+##### CRD
+
+The cloud component will define a dedicated CRD that the network plugin will use
+to issue cloud egress IP requests. The network plugin will in-turn finalize its
+setup and set its final egress IP status pending the CR's status indicating a
+successful/unsuccessful cloud assignment. For example: if the cloud assignment
+should be unsuccessful; it will be up to the network plugin to notify the user
+of this, by setting its own status field for the egress IP assignment
+accordingly.
+
+The CRD will have `scope: Cluster` (since egress IP is cluster-scoped and node
+private IP assignment have no relation to namespaces), always have a single and
+unique CR, and define the following format:
+
+```yaml
+apiVersion: cloud.openshift.io/v1
+kind: CloudPrivateIPConfig
+metadata:
+  name: 192.168.126.11
+spec:
+  node: nodeX
+  status:
+    node: node X
+    conditions:
+    - message: ""
+      reason: ""
+      status: "True|False|Pending"
```

Comment:
condition statuses are almost always `v1.ConditionStatus`, which is `True`/`False`/`Unknown`. If you want a "Pending" state, that could be a separate status

Reply:
Yes, "Pending" was incorrect on my part, it's "Unknown" which will be accompanied with a `reason` stating something along the lines of: "CloudResponsePending"